src/field_theory/tower.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import ring_theory.algebra_tower
+import linear_algebra.finite_dimensional
+
+/-! # Tower of field extensions
+Tower law for arbitrary extensions and finite extensions. -/
+
+universes u v w u₁ v₁ w₁
+open_locale classical big_operators
+
+section semiring
+
+variables {R : Type u} {S : Type v} {A : Type w}
+variables [comm_semiring R] [comm_semiring S] [semiring A]
+variables [algebra R S] [algebra S A] [algebra R A] [is_algebra_tower R S A]
+
+namespace submodule
+
+open is_algebra_tower
+
+theorem smul_mem_span_smul_of_mem {s : set S} {t : set A} {k : S} (hks : k ∈ span R s)
+  {x : A} (hx : x ∈ t) : k • x ∈ span R (s • t) :=
```

Comment:
I this and the next few lemmas specific to towers? They look like they should be moved to earlier files.

Reply:
tbh the whole import hierarchy is a bit messy. maybe algebera_tower should occupy a very low position in the import hierarchy (i.e. imported by a lot of files) and this file should not exist at all.

Reply:
I wanted to put the lemmas in this file because they use the `span R s = ⊤` assumption, so they're directly related to the spanning part of `is_basis.smul`.

Reply:
I think I agree with Patrick.
If the import hierarchy is messy, we better clean it up, before adding stuff and (potentially) making it harder to untangle later on.

Reply:
So if you have an idea how to improve the import situation, I would be really happy with a refactor PR after this one. What do you think?

Reply:
I don't know; have you any idea?

Reply:
If you agree, I suggest that we merge this like it is, but then take a step back to refactor things a bit, and untangle the import hierarchy.

Reply:
sure

Reply:
This conversation sounded like an agreement to punt on this one, with everyone agreeing that they don't really feel like refactoring. Can't we just move these lemmas earlier, as @PatrickMassot suggested?

Reply:
Presumably those lemmas should just be in `ring_theory/algebra_tower`.

Reply:
Okay, I just did it myself.

I also added a module-doc to `ring_theory/algebra_tower`, and explained the division of the material between the two files. @kckennylau, could you check if I said anything stupid?